File: src/gazebo_mocap_node.cpp
#include <map>
#include <string>
#include <sstream>

#include <ros/ros.h>
#include <tf2_ros/transform_broadcaster.h>
#include <gazebo_msgs/ModelStates.h>
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/TransformStamped.h>

class GazeboMocapNode {

    public:
        GazeboMocapNode(ros::NodeHandle, ros::NodeHandle);
        void run();

    private:
        ros::NodeHandle nh, pnh;
        tf2_ros::TransformBroadcaster broadcaster;
        void model_states_callback(const gazebo_msgs::ModelStates::ConstPtr&);
        std::map<std::string, ros::Publisher> publishers;
        ros::Subscriber model_states_sub;

};

GazeboMocapNode::GazeboMocapNode(ros::NodeHandle nh, ros::NodeHandle pnh) {
    this->nh = nh;
    this->pnh = pnh;
    this->model_states_sub = this->nh.subscribe("/gazebo/model_states", 1, &GazeboMocapNode::model_states_callback, this);

}

void GazeboMocapNode::model_states_callback(const gazebo_msgs::ModelStates::ConstPtr& model_states_msg) {

    for (std::size_t i = 0; i < model_states_msg->name.size(); i++) {
        std::string name = model_states_msg->name[i];
        // Check if "drone" is contained in the model state name
        if (name.find("drone") != std::string::npos) {
            // If there is no publisher yet, create it
            if (publishers.find(name.c_str()) == publishers.end()) {
                std::stringstream ss;
                ss << "/" << name << "/mavros/vision_pose/pose";
                std::string topic = ss.str();
                publishers[name] = nh.advertise<geometry_msgs::PoseStamped>(topic, 1);
            } else {

                // Get Gazebo model pose
                geometry_msgs::Pose model_pose = model_states_msg->pose[i];

                // Publish pose
                geometry_msgs::PoseStamped pose_msg;
                pose_msg.header.frame_id = "world";
                pose_msg.header.stamp = ros::Time::now();
                pose_msg.pose = model_pose;

                publishers[name].publish(pose_msg);

                std::stringstream ss;
                ss << name << "/base_link";

                // Broadcast transform
                geometry_msgs::TransformStamped transform_msg;
                transform_msg.header.frame_id = "world";
                transform_msg.header.stamp = ros::Time::now();
                transform_msg.child_frame_id = ss.str();

                transform_msg.transform.translation.x = model_pose.position.x;
                transform_msg.transform.translation.y = model_pose.position.y;
                transform_msg.transform.translation.z = model_pose.position.z;

                transform_msg.transform.rotation = model_pose.orientation;

                broadcaster.sendTransform(transform_msg);

            }

        }
        // ROS_INFO_STREAM("Name: " << i);
    }

}

void GazeboMocapNode::run() {

    ros::Rate rate(100.0);

    while (ros::ok()) {
        ros::spinOnce();
        rate.sleep();
    }

}

int main(int argc, char** argv) {
    ros::init(argc, argv, "gazebo_mocap_node");
    ros::NodeHandle nh(""), pnh("~");
    GazeboMocapNode gazebo_mocap_node(nh, pnh);
    gazebo_mocap_node.run();
    return 0;
}

File: src/gazebo_mocap_node.py
#!/usr/bin/env python3
"""Relays Gazebo model states as MAVROS vision pose messages."""
from __future__ import absolute_import, division, print_function

import sys

import rospy
import tf2_ros
from gazebo_msgs.msg import ModelStates
from geometry_msgs.msg import PoseStamped, TransformStamped


class GazeboMocapNode:

    def __init__(self):

        # Initialize node
        rospy.init_node('gazebo_mocap_node', argv=sys.argv)

        self.num_agents = rospy.get_param('~num_agents')
        self.vehicle = 'drone'

        self.broadcaster = tf2_ros.TransformBroadcaster()

        # Vision pose publishers
        self.vision_pose_pubs = []
        for i in range(self.num_agents):
            topic = '/{}_{}/mavros/vision_pose/pose'.format(self.vehicle, i + 1)
            vision_pose_pub = rospy.Publisher(topic, PoseStamped, queue_size=1)
            self.vision_pose_pubs.append(vision_pose_pub)

        # Model states subscriber
        self.model_states_msg = None
        self.model_states_sub = rospy.Subscriber('/gazebo/model_states',
                                                 ModelStates,
                                                 callback=self.model_states_callback,
                                                 queue_size=1)

    def send_transform(self):

        # Wait to receive model states message before publishing
        if self.model_states_msg is None:
            return

        # Convert Gazebo model states into pose and twist
        for i, name in enumerate(self.model_states_msg.name):
            if name.startswith(self.vehicle):
                index = int(name.split('_')[-1]) - 1
                pose = PoseStamped()
                pose.header.stamp = rospy.Time.now()
                pose.header.frame_id = 'world'
                pose.pose = self.model_states_msg.pose[i]
                self.vision_pose_pubs[index].publish(pose)
                child_frame = '{}_{}'.format(self.vehicle, index + 1)

                transform = TransformStamped()

                transform.header.stamp = rospy.Time.now()
                transform.header.frame_id = 'world'
                transform.child_frame_id = child_frame

                p = pose.pose

                transform.transform.translation.x = p.position.x
                transform.transform.translation.y = p.position.y
                transform.transform.translation.z = p.position.z

                transform.transform.rotation.x = p.orientation.x
                transform.transform.rotation.y = p.orientation.y
                transform.transform.rotation.z = p.orientation.z
                transform.transform.rotation.w = p.orientation.w
                self.broadcaster.sendTransform(transform)

    def model_states_callback(self, model_states_msg):
        self.model_states_msg = model_states_msg

    def run(self):
        rate_param = rospy.get_param('~rate', default=100)
        rate = rospy.Rate(rate_param)
        while not rospy.is_shutdown():
            self.send_transform()
            try:
                rate.sleep()
            except rospy.ROSInterruptException:
                pass


def main():
    GazeboMocapNode().run()


if __name__ == '__main__':
    main()
